src/routes/transactions/transactions.service.ts

```diff
@@ -175,4 +181,125 @@ export class TransactionsService {
       results,
     };
   }
+
+  async getTransactionQueue(
+    chainId: string,
+    routeUrl: Readonly<URL>,
+    safeAddress: string,
+    paginationData: PaginationData,
+  ): Promise<Page<QueuedItem>> {
+    const pagination = this.getAdjustedPagination(paginationData);
+    const safeInfo = await this.safeRepository.getSafe(chainId, safeAddress);
+    const domainTransactions =
+      await this.safeRepository.getTransactionQueueByNonce(
+        chainId,
+        safeAddress,
+        pagination.limit,
+        pagination.offset,
+      );
+
+    const transactions = await Promise.all(
+      this.adjustTransactionsPage(domainTransactions).results.map(
+        async (transaction) =>
+          this.multisigTransactionMapper.mapTransaction(
+            chainId,
+            transaction,
+            safeInfo,
+          ),
+      ),
+    );
+
+    const results = await this.queuedItemsMapper.getQueuedItems(
+      transactions,
+      safeInfo,
+      this.getPreviousPageLastNonce(domainTransactions, paginationData),
+      this.getNextPageFirstNonce(domainTransactions),
+    );
+
+    const { next, previous } = domainTransactions;
+    const nextURL = cursorUrlFromLimitAndOffset(routeUrl, next);
+    const previousURL = cursorUrlFromLimitAndOffset(routeUrl, previous);
+
+    return {
+      count: results.length,
+      next: nextURL?.toString() ?? null,
+      previous: previousURL?.toString() ?? null,
+      results,
+    };
+  }
+
+  /**
+   * Adjusts the pagination data to return extra items in both "edges" of the current page:
+   * - If no pagination data info, then return the original pagination data.
+   * - If it is the first page (offset 0), then return offset: 0, limit: limit + 1.
+   * - If it is not the first page, then return offset: offset - 1, limit: limit + 2.
+   * @param paginationData pagination data to adjust.
+   * @returns pagination data adjusted.
+   */
+  private getAdjustedPagination(
+    paginationData: PaginationData,
+  ): PaginationData {
+    if (!paginationData.limit || !paginationData.offset) {
+      return paginationData;
+    }
+    if (paginationData.offset === 0) {
+      return new PaginationData(
+        paginationData.limit + 1,
+        paginationData.offset,
+      );
+    } else {
+      return new PaginationData(
+        paginationData.limit + 2,
+        paginationData.offset - 1,
+      );
+    }
+  }
```

Comment:
@hectorgomezv why this function is necessary?

Reply:
> Or what's the purpose of adjusting the pagination values?  
The second one :)

Reply:
I just followed the Rust implementation, in which, an extra item is needed on both the previous and the next page of the page being requested. This is necessary because some logic is in place to help the clients draw the view. The client shows transaction groups and needs to know whether the first/last element is part of a group of the previous/next page and other labeling-related stuff. You can find more details in the Rust service Swagger (and in the code itself of course).

Having that said, since one extra item is requested on each _side_ of the page, an adjustment is needed to remove the extra items from the response. (But again, here I'm just following the logic implemented for the Rust service, IMHO all this logic eventually could be placed in the client side, to have a more agnostic backend response 🤔)